From 18e5973e551fab8cde03567edc8041707f67e628 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:49 -0400
Subject: [ethernet] mlx4: Fix errors in MAC address changing when port is down

Message-id: <17ca755ff2118e3a697ef4995bf7de4d98a0d169.1410425016.git.avadai@redhat.com>
Patchwork-id: 93716
O-Subject: [Patch RHEL7.1 V1 113/197] net/mlx4_en: Fix errors in MAC address changing when port is down
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Shani Michaelli <shanim@mellanox.com>

This patch fix an issue that happen when changing the MAC address when
the port is down, described as follows:
1. Set the port down.
2. Change the MAC address - mlx4_en_set_mac() will change dev->dev_addr.
3. Set the port up - will result in mlx4_en_do_uc_filter that will
   remove the prev_mac entry from the mac_hash db.
4. Changing the MAC address again will eventually trigger the call to
   mlx4_en_replace_mac() in order to replace prev_mac with dev_addr but
   the prev_mac entry is already not exist in the mac_hash db therefore
   the operation fails.

The fix is to set the prev_mac with the new MAC address so in step 3
above, after setting the port up mlx4_en_get_qp() is updating the
mac_hash with the entry of dev_addr which is equal to prev_mac.
Therefore in step 4, when calling mlx4_en_replace_mac, the entry related
to prev_mac exist in mac_hash and the replace operation succeed.

Reviewed-by: Eyal Perry <eyalpe@mellanox.com>
Signed-off-by: Shani Michaeli <shanim@mellanox.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit ee755324a3aa1459c7e184e12c170b89bd159f22)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
index fba3c8e..eea7868 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_netdev.c
@@ -770,11 +770,12 @@ static int mlx4_en_do_set_mac(struct mlx4_en_priv *priv)
        priv->dev->dev_addr, priv->prev_mac);
   if (err)
    en_err(priv, "Failed changing HW MAC address\n");
-  memcpy(priv->prev_mac, priv->dev->dev_addr,
-         sizeof(priv->prev_mac));
  } else
   en_dbg(HW, priv, "Port is down while registering mac, exiting...\n");
 
+ memcpy(priv->prev_mac, priv->dev->dev_addr,
+        sizeof(priv->prev_mac));
+
  return err;
 }
 
-- 
1.7.1